### PR TITLE
Issue #153 Add codefix for S2934 (property assignment of a generic readonly field)

### LINKED ISSUE #153: Add codefix for S2934 (property assignment of a generic readonly field)

### PR DESCRIPTION
**Pull request by tvsonar**
Fixes #153

### CODE REVIEW

File: src/SonarLint/Rules/GenericReadonlyFieldPropertyAssignmentCodeFixProvider.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace SonarLint.Rules
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp)]
+    public class GenericReadonlyFieldPropertyAssignmentCodeFixProvider : CodeFixProvider
+    {
+        internal const string TitleRemove = "Remove assignment";
```
Comment (dbolkensteyn) [inline src/SonarLint/Rules/GenericReadonlyFieldPropertyAssignmentCodeFixProvider.cs]: @tvsonar i wouldn't suggest this code fix - it is likely to break code

File: src/SonarLint/Rules/GenericReadonlyFieldPropertyAssignmentCodeFixProvider.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        internal const string TitleAddClassConstraint = "Add reference type constraint";
```
Comment (dbolkensteyn) [inline src/SonarLint/Rules/GenericReadonlyFieldPropertyAssignmentCodeFixProvider.cs]: @tvsonar this seems like the better one - but why isn't it being offered?

File: src/Tests/SonarLint.UnitTest/TestCases/GenericReadonlyFieldPropertyAssignment.AddConstraint.Fixed.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Tests.TestCases
```
Comment (duncanpMS) [inline src/Tests/SonarLint.UnitTest/TestCases/GenericReadonlyFieldPropertyAssignment.AddConstraint.Fixed.cs]: There's an obscure corner case with multiple generic arguments that seems to generate a false positive. I think it's too obscure to worry about, though.

    class P2<A, B> where A : class, IPoint where B : A
    {
        readonly A pointA;
        readonly B pointB;

        public P2(A a, B b)
        {
            this.pointA = a;
            this.pointB = b;
        }

        public void Add(int i)
        {
            pointA.X += i;
            pointB.X += i + i;   // diagnostic fires here
        }
    }
`

File: src/Tests/SonarLint.UnitTest/TestCases/GenericReadonlyFieldPropertyAssignment.AddConstraint.Fixed.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            this.point = point;
+            this.point.X = 1; // Compliant, we are in the constructor
+        }
+
+        public void MovePointVertically(int newX)
+        {
+            point.X = newX; //Noncompliant; if point is a struct, then nothing happened
+            point.X++; //Noncompliant; if point is a struct, then nothing happened
+            Console.WriteLine(point.X);
```
Comment (duncanpMS) [inline src/Tests/SonarLint.UnitTest/TestCases/GenericReadonlyFieldPropertyAssignment.AddConstraint.Fixed.cs]: Another corner case (against, not very common):
            int i = point.X = newX;
            point.X = i = newX;
The "remove" constraint strips out the whole line so the assignment to *i* is lost.

**Comment by duncanpMS** [comment]
LGTM

File: src/Tests/SonarLint.UnitTest/TestCases/GenericReadonlyFieldPropertyAssignment.AddConstraint.Fixed.cs (hunk 3 above)
Comment (tvsonar) [inline src/Tests/SonarLint.UnitTest/TestCases/GenericReadonlyFieldPropertyAssignment.AddConstraint.Fixed.cs]: Created #459 to cover this.

File: src/Tests/SonarLint.UnitTest/TestCases/GenericReadonlyFieldPropertyAssignment.AddConstraint.Fixed.cs (hunk 4 above)
Comment (tvsonar) [inline src/Tests/SonarLint.UnitTest/TestCases/GenericReadonlyFieldPropertyAssignment.AddConstraint.Fixed.cs]: Fixed this by only proposing the fix if the assignment is directly in a statement. Created #460 for the proper assignment removal.
